### PR TITLE
Show tasks

### PR DESCRIPTION
**Pull request by mim24sd**
In this pull request I want to:
1. show tasks.
2. sort tasks by time.
3. show the number of all tasks and uncompleted tasks.
4. show the creation date as a tooltip on each task.

### CODE REVIEW

File: assests/javascript/index.js [hunk 1]
```diff
@@ -5,23 +5,30 @@ const addTaskIcon = document.getElementById("add-task-icon");
...
 const networkErrorMassage = document.getElementById("network-error");
+const taskList = document.getElementById("task-list");
+const filterTasksView = document.getElementById("filter-tasks");
 
 addTaskInput.addEventListener(
   "keydown",
   (event) => {
     if (event.key === "Enter") {
       handleNewTask(addTaskInput.value);
+      addTaskInput.value = "";
```
Comment (blacksrc) [inline assests/javascript/index.js]: Now it's time to make a function for it.

File: assests/javascript/index.js [hunk 2]
```diff
@@ -5,23 +5,30 @@ const addTaskIcon = document.getElementById("add-task-icon");
...
 });
 
-async function handleNewTask(task) {
+showAllTasks();
+
+async function handleNewTask(newTask) {
   hideErrors();
-  addTask(task);
+  await addTask(newTask);
+  showAllTasks();
```
Comment (blacksrc) [inline assests/javascript/index.js]: It's not necessary to re-fetch all the tasks again in case that new task fails. right?

File: assests/javascript/index.js [hunk 3]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+    const response = await fetch(`${baseUrl}/todos`);
+    const {data} = await response.json();
+
+    return sortTasks(data);
+  } catch {
+    showConnecionErrorMassage();
+  }
+}
+
+async function showAllTasks() {
```
Comment (blacksrc) [inline assests/javascript/index.js:65]:
```suggestion
async function renderTasks() {
```

File: assests/javascript/index.js [hunk 4]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+  const uncompletedTasks = getUncompletedTasks(tasks);
+
+  filterTasksView.innerHTML = 
+    `<li><a>All Tasks ${tasks.length}</a></li>
+    <li><a>Completed ${completedTasks.length}</a></li>
+    <li><a>Uncompleted ${uncompletedTasks.length}</a></li>`
+}
+
+function getCompletedTasks(tasks) {
+  return tasks.filter((task) => {return task.isDone === true;});
```
Comment (blacksrc) [inline assests/javascript/index.js]:
```suggestion
  return tasks.filter((task) =>  task.isDone);
```

Comment (blacksrc) [reply assests/javascript/index.js]: Can do the same for the others.

File: assests/javascript/index.js [hunk 5]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+  return tasks.filter((task) => {return task.isDone === false;});
+}
+
+function sortTasks(tasks) {
+  return tasks.sort((task1, task2) => new Date(task2.createdAt).getTime() - new Date(task1.createdAt).getTime());
+}
+
+function isChecked(isDone) {
+  return isDone === true;
+}
```
Comment (blacksrc) [inline assests/javascript/index.js]: This function is redundant.

File: assests/javascript/index.js [hunk 6]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+
+async function filterTasks(tasks) {
+  const completedTasks = getCompletedTasks(tasks);
+  const uncompletedTasks = getUncompletedTasks(tasks);
+
+  filterTasksView.innerHTML = 
+    `<li><a>All Tasks ${tasks.length}</a></li>
+    <li><a>Completed ${completedTasks.length}</a></li>
+    <li><a>Uncompleted ${uncompletedTasks.length}</a></li>`
```
Comment (blacksrc) [inline assests/javascript/index.js]: Move back all `li` and `a` and the text to the DOM. Add an extra span with the `id` and fill the numbers only.

File: assests/javascript/index.js [hunk 7]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+              <i class="fa-solid fa-pen-to-square edit-task-icon"></i>
+              <i class="fa-solid fa-xmark delete-task-icon"></i>
+
+              <p class="tooltiptext">${convertTime(task.createdAt)}</p>
+            </li>`;}).join("");
+
+  filterTasks(tasks);
+}
+
+function convertTime(time) {
```
Comment (blacksrc) [inline assests/javascript/index.js]:
```suggestion
function formatDate(time) {
```

File: assests/javascript/index.js [hunk 8]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+  const dateTime = new Date(time);
+  const year = dateTime.toLocaleDateString("en-US", { year: "numeric" });
+  const month = dateTime.toLocaleDateString("en-US", { month: "short" });
+  const date = dateTime.toLocaleDateString("en-US", { day: "numeric" });
```
Comment (blacksrc) [inline assests/javascript/index.js]: Fix the variable name.

File: assests/javascript/index.js [hunk 9]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+
+  return `${date} ${month} ${year}`;
+}
+
+async function filterTasks(tasks) {
```
Comment (blacksrc) [inline assests/javascript/index.js]: Naming is not good. The function is not filtering anything. Think about this a bit more.

File: assests/javascript/index.js [hunk 10]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+    return sortTasks(data);
+  } catch {
+    showConnecionErrorMassage();
+  }
+}
+
+async function showAllTasks() {
+  const tasks = await getAllTasks();
+
+  taskList.innerHTML = tasks.map((task) => {
```
Comment (blacksrc) [inline assests/javascript/index.js:71]: Destructure `task` while mapping.

File: assests/javascript/index.js [hunk 11]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
   }
 }
 
+async function getAllTasks() {
+  try {
+    const response = await fetch(`${baseUrl}/todos`);
+    const {data} = await response.json();
+
+    return sortTasks(data);
```
Comment (blacksrc) [inline assests/javascript/index.js]: This function should not care about sorting. This is better to be called inside `showAllTasks` instead.

File: assests/javascript/index.js [hunk 12]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+
+function getCompletedTasks(tasks) {
+  return tasks.filter((task) => {return task.isDone === true;});
+}
+
+function getUncompletedTasks(tasks) {
+  return tasks.filter((task) => {return task.isDone === false;});
+}
+
+function sortTasks(tasks) {
```
Comment (blacksrc) [inline assests/javascript/index.js]: Think about this function's naming a bit. `sortTasks` is a bit generic,

File: assests/javascript/index.js [hunk 13]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+  } catch {
+    showConnecionErrorMassage();
+  }
+}
+
+async function showAllTasks() {
+  const tasks = await getAllTasks();
+
+  taskList.innerHTML = tasks.map((task) => {
+    return `<li class="task-box tooltip" id=${task.id}>
```
Comment (blacksrc) [inline assests/javascript/index.js]: id value is not inside `""`. Check the DOM after creating a new tasks.

File: assests/javascript/index.js [hunk 14]
```diff
@@ -44,6 +51,68 @@ async function addTask(task) {
...
+              <input type="checkbox" class="task-check-box" ${isChecked(task.isDone)}></input>
+              <p class="task-title">${task.text}</p>
+              <i class="fa-solid fa-pen-to-square edit-task-icon"></i>
+              <i class="fa-solid fa-xmark delete-task-icon"></i>
+
+              <p class="tooltiptext">${convertTime(task.createdAt)}</p>
+            </li>`;}).join("");
```
Comment (blacksrc) [inline assests/javascript/index.js]: I'm not sure if editor is formatting your code correctly or not. But a new line after `;` would be nice.

Comment (mim24sd) [reply assests/javascript/index.js]: done

File: assests/javascript/index.js (hunk 13 above)
Comment (mim24sd) [reply assests/javascript/index.js]: yesss fixed

File: assests/javascript/index.js (hunk 12 above)
Comment (mim24sd) [reply assests/javascript/index.js]: I replaced it by `sortTasksByTime`

File: assests/javascript/index.js (hunk 11 above)
Comment (mim24sd) [reply assests/javascript/index.js]: ok

File: assests/javascript/index.js (hunk 9 above)
Comment (mim24sd) [reply assests/javascript/index.js]: replaced by `classifyTasks`

File: assests/javascript/index.js (hunk 8 above)
Comment (mim24sd) [reply assests/javascript/index.js]: replaced by `day`

File: assests/javascript/index.js (hunk 7 above)
Comment (mim24sd) [reply assests/javascript/index.js]: done

File: assests/javascript/index.js (hunk 1 above)
Comment (mim24sd) [reply assests/javascript/index.js]: done

File: assests/javascript/index.js (hunk 2 above)
Comment (mim24sd) [reply assests/javascript/index.js]: yessss thanks

File: assests/javascript/index.js (hunk 4 above)
Comment (mim24sd) [reply assests/javascript/index.js]: oh yessss
that was completely a mistake

File: assests/javascript/index.js (hunk 5 above)
Comment (mim24sd) [reply assests/javascript/index.js]: yeeees

File: assests/javascript/index.js (hunk 10 above)
Comment (mim24sd) [reply assests/javascript/index.js:71]: such a great idea

File: assests/javascript/index.js (hunk 6 above)
Comment (mim24sd) [reply assests/javascript/index.js]: okkkk thanks

File: assests/javascript/index.js (hunk 10 above)
Comment (blacksrc) [reply assests/javascript/index.js:71]: Nice. Can be better:
Google: Destructuring objects as function parameters in ES6

File: assests/javascript/index.js [hunk 15]
```diff
@@ -34,16 +42,87 @@ async function addTask(task) {
...
+}
+
+async function classifyTasks(tasks) {
+  const completedTasks = getCompletedTasks(tasks);
+  const uncompletedTasks = getUncompletedTasks(tasks);
+
+  numberOfAllTasks.innerText = tasks.length;
+  numberOfCompletedTasks.innerText = completedTasks.length;
+  numberOfUncompletedTasks.innerText = uncompletedTasks.length;
+}
```
Comment (blacksrc) [inline assests/javascript/index.js]: Beautiful! Clean, Readable 🤩

File: assests/javascript/index.js [hunk 16]
```diff
@@ -34,16 +42,87 @@ async function addTask(task) {
...
+
+function getCompletedTasks(tasks) {
+  return tasks.filter((task) => { task.isDone });
```
Comment (blacksrc) [inline assests/javascript/index.js]: Check my previous comment carefully and explain:
1. Why this wont work.
2. How should we change it and why the change would work.
(Explain in a voice message)

File: assests/javascript/index.js [hunk 17]
```diff
@@ -34,16 +42,87 @@ async function addTask(task) {
...
+}
+
+function getUncompletedTasks(tasks) {
+  return tasks.filter((task) => { return task.isDone === false; });
```
Comment (blacksrc) [inline assests/javascript/index.js]: Change this part based on the other [comment above](https://github.com/mim24sd/todo/pull/3/files#r1025592543)

File: assests/javascript/index.js [hunk 18]
```diff
@@ -34,16 +42,87 @@ async function addTask(task) {
...
+}
+
+function sortTasksByTime(tasks) {
+  return tasks.sort((task1, task2) => new Date(task2.createdAt).getTime() - new Date(task1.createdAt).getTime());
+}
+
+function emptyNewTask() {
```
Comment (blacksrc) [inline assests/javascript/index.js]:
```suggestion
function clearNewTaskInput() {
```

File: index.html [hunk 19]
```diff
@@ -1,63 +1,42 @@
...
+    <h1 class="title">Todo</h1>
+    <div id="network-error" class="network-error-box">
+      <p>Something went wrong.</p>
+      <p>Check your connection or try again later.</p>
+    </div>
+    <nav>
+      <ul class="nav-list-box">
+        <li><a>All Tasks <span id="number-of-all-tasks"></span></a></li>
+        <li><a>Completed <span id="number-of-completed-tasks"></span></a></li>
+        <li><a>Uncompleted <span id="number-of-uncompleted-tasks"></span></a></li>
```
Comment (blacksrc) [inline index.html]: Nice ⚡ Just a naming suggestion:
```suggestion
        <li><a>All Tasks <span id="all-tasks-count"></span></a></li>
        <li><a>Completed <span id="completed-tasks-count"></span></a></li>
        <li><a>Uncompleted <span id="uncompleted-tasks-count"></span></a></li>
```

Comment (mim24sd) [reply index.html]: thanks done

File: assests/javascript/index.js (hunk 18 above)
Comment (mim24sd) [reply assests/javascript/index.js]: done

File: assests/javascript/index.js (hunk 16 above)
Comment (mim24sd) [reply assests/javascript/index.js]: doneeee

File: assests/javascript/index.js (hunk 15 above)
Comment (mim24sd) [reply assests/javascript/index.js]: 💫💫💫

File: assests/javascript/index.js (hunk 17 above)
Comment (mim24sd) [reply assests/javascript/index.js:110]: doneeee

**Review by blacksrc — APPROVED** [review]

File: index.html (hunk 19 above)
Comment (mim24sd) [reply index.html]: done

File: assests/javascript/index.js (hunk 18 above)
Comment (mim24sd) [reply assests/javascript/index.js]: day neee thanks

File: assests/javascript/index.js (hunk 17 above)
Comment (mim24sd) [reply assests/javascript/index.js]: yesss

Comment (mim24sd) [reply assests/javascript/index.js]: done
